Integration of functional groups identification functionality following the Ertl algorithm

JonasSchaub opened the pull request:
Reference issue: #854

JonasSchaub commented:
@johnmay and @egonw ,

I have integrated the FunctionalGroupsFinder (dropped the "Ertl" as suggested by John) into the cdk-fragment module. 

I had to change the dependencies of the module a bit, could you tell me whether this is ok? 
And do you think I should implement IFragmenter in FunctionalGroupsFinder? This should not be complicated but I think using it through the interface would mean the loss of a lot of functionality for the class.

johnmay commented on tool/fragment/pom.xml:
```diff
@@ -63,7 +63,6 @@
             <groupId>${project.groupId}</groupId>
             <artifactId>cdk-data</artifactId>
             <version>${project.parent.version}</version>
-            <scope>test</scope>
```
cdk-data should not be in the non-test scope. Generally you may need a builder passed in you get it from the provided molecule/container.

JonasSchaub replied:
I used "instanceof PseudoAtom" to detect pseudo atom instances. This was my only non-test use of cdk-data. I've adjusted this in my last commit (https://github.com/cdk/cdk/pull/1039/commits/a19085ef204925cbf8135a0498094b16cb3fecd0) and re-assigned the test scope to cdk-data.

johnmay commented on tool/fragment/src/main/java/org/openscience/cdk/fragment/FunctionalGroupsFinder.java:
```diff
@@ -0,0 +1,1339 @@
+     * Atoms of these elements are exclusively accepted in the input molecule if(!) the strict input restrictions are
+     * activated (turned off by default).
+     */
+    public static final Set<Integer> NONMETAL_ATOMIC_NUMBERS = new HashSet<>(Arrays.asList(1, 2, 6, 7, 8, 9, 10, 15, 16, 17, 18, 34, 35, 36, 53, 54, 86));
```
IAtom.Hydrogen, IAtom.Helium, etc..

JonasSchaub replied:
Corrected in https://github.com/cdk/cdk/pull/1039/commits/8bde105371a29bb0a6fa439619652c41c2d2b5c0

johnmay commented on tool/fragment/src/main/java/org/openscience/cdk/fragment/FunctionalGroupsFinder.java:
```diff
@@ -0,0 +1,1339 @@
+            }
+        }
+        this.bondMapCache = EdgeToBondMap.withSpaceFor(tmpMolecule);
+        this.adjListCache = GraphUtil.toAdjList(tmpMolecule, this.bondMapCache);
```
You probably don't need this anymore since AtomContainer2 provides these optimisations... but not wrong do this.

johnmay reviewed (commented):
Looks good, few minor things to tweak then will deep dive a bit more. Overall I think I would make the main API as follows:

```
// fill in the fgs[] array 1..num_fgps, 0 means not in a function group;
// return: the number of functional groups
public int markFunctionalGroups(int[] fgs, IAtomContain mol) {
 // ...
}

public List<IAtomContainer> extractFunctionalGroups(IAtomContain mol)
{
  int[] fgs = new int[mol.getAtomContainer()];
  int cnt = markFunctionalGroups(fgs, mol);
  
  List<IAtomContainer> res = new ArrayList<>();
  for (int grp=1; grp<=cnt; grp++) {
    IAtomContainer part = // newMol
    // plus extra C/Car atoms, note this does a deep copy, but a shallow copy would perhaps also be reasonable
    AtomContainerManpipulator.copy(part, mol, a -> a.getIndex());
    res.add(part);
  }
  return res;
}
```

johnmay commented:
Overall patch looks good and with a few minor changes am happy to merge it in. The biggest blockers are the cdk-data as a dependency and too many public methods. 

As a minor improvement I used to use the GraphUtil.toAdjList a lot in older code but nowadays just using the IAtomContainer should be as fast - see https://github.com/cdk/cdk/wiki/AtomContainer2. You don't need to do anything for this as it the default if you come from the builders. There are still some gotchas but overall for this it should work well.

JonasSchaub commented on tool/fragment/src/main/java/org/openscience/cdk/fragment/FunctionalGroupsFinder.java:
```diff
@@ -0,0 +1,1339 @@
+                    continue;
+                }
+                if (Objects.isNull(tmpConnectedAtom.getImplicitHydrogenCount())) {
+                    tmpConnectedAtom.setImplicitHydrogenCount(1);
```
Adjusted in https://github.com/cdk/cdk/pull/1039/commits/644f0a5da4cf9bf17b88984f9f5621cee8d73638

JonasSchaub commented on tool/fragment/src/main/java/org/openscience/cdk/fragment/FunctionalGroupsFinder.java:
```diff
@@ -0,0 +1,1339 @@
+     * @throws NullPointerException if any parameter is null
+     * @throws IllegalArgumentException if the input molecule causes any other type of exception while processing
+     */
+    public static void applyPreprocessing(IAtomContainer aMolecule, Aromaticity anAromaticityModel) throws NullPointerException, IllegalArgumentException {
```
Done in https://github.com/cdk/cdk/pull/1039/commits/c02aa4c0c6d9742eadbd227ff586f64f586cc3f0

JonasSchaub commented on tool/fragment/src/main/java/org/openscience/cdk/fragment/FunctionalGroupsFinder.java:
```diff
@@ -1191,7 +1190,8 @@ private void expandEnvironmentGeneralized(IAtom aFunctionalGroupAtom, IAtomConta
      */
     private boolean isHeteroatom(IAtom anAtom) {
         Integer tmpAtomicNr = anAtom.getAtomicNumber();
-        return tmpAtomicNr != 1 && tmpAtomicNr != 6 && tmpAtomicNr != 0 && tmpAtomicNr != null && !(anAtom instanceof PseudoAtom);
```
Thanks for the hint with the interface! I have added this in https://github.com/cdk/cdk/pull/1039/commits/c9ef1dc26ef1188715efe9681dc0ffa21edf1da8. Now, I think I'm really testing for pseudo atoms in every possible way. Please let me know should some of these conditions be unnecessary.

JonasSchaub replied:
I think your comment here https://github.com/cdk/cdk/pull/1040#issuecomment-1946104509 got posted in the wrong PR. Sorry for creating confusion with my two parallel PRs.

JonasSchaub commented:
@johnmay,
Thank you for your comments!

cdk-data is now out of the non-test scope again and I have implemented some of your suggestions.

About the many public methods: I have envisioned most of them as utilities for the user. This way, they can be sure that their input molecule is valid or not without simply trying it out and causing an exception. For example, testing whether a molecule contains metal, metalloid, or pseudo atoms is not completely straightforward and therefore I think utilities like this can be very helpful. What do you think? If you still think they are unnecessary, I will remove what I don't actually use internally and make the rest package private as you suggested.

About your suggested main API: I have considered it but implementing the integer array return type mapping atom indices to functional group indices would be non-straightforward and too confusing to the user in the end, I think. I could implement this for the atoms that are marked according to the algorithm but on top of that, the Ertl algorithm also adds pseudo atoms and explicit hydrogen atoms as functional group environments which could not be represented nicely in this return type. The connected environmental carbon atoms we attach to the functional groups are also not the original carbon atoms from the molecule but substitutes that mainly contain the information whether the environmental carbon is aliphatic or aromatic.
As for your suggestion "public List<IAtomContainer> extractFunctionalGroups(IAtomContain mol)", I have renamed the central find() method to this name (https://github.com/cdk/cdk/pull/1039/commits/e6653ba7bf797dce0fe36c609ca6a62c87899c71). The method already had the same parameter and return type signature.

And about adjacency list vs atom container 2 (sorry for the long text!): thank you for pointing out that AtomContainer2 should be as fast as our solution now. I will keep this in mind. But please forgive me that I am unwilling to change this here now since this is not my code originally (I only took over the responsibility for it) and I don't want to change the entire internal implementation at this step anymore. Sorry!

johnmay commented:
> @johnmay, Thank you for your comments!
> 
> cdk-data is now out of the non-test scope again and I have implemented some of your suggestions.
> 
> About the many public methods: I have envisioned most of them as utilities for the user. This way, they can be sure that their input molecule is valid or not without simply trying it out and causing an exception. For example, testing whether a molecule contains metal, metalloid, or pseudo atoms is not completely straightforward and therefore I think utilities like this can be very helpful. What do you think? If you still think they are unnecessary, I will remove what I don't actually use internally and make the rest package private as you suggested.

Utilities yes - but in the right place. I understand the CDK API is quite bloated so it can be hard to find the right thing but adding more utility methods to places that someone would not think to look doesn't help IMO. For example we already have an isMetal as part of the Elements enum class: https://github.com/cdk/cdk/blob/main/base/core/src/main/java/org/openscience/cdk/config/Elements.java#L516C27-L516C34

Would someone looking for an isMetal() etc function expect to find it on this functional group extraction algorithm?

> 
> About your suggested main API: I have considered it but implementing the integer array return type mapping atom indices to functional group indices would be non-straightforward and too confusing to the user in the end, I think. I could implement this for the atoms that are marked according to the algorithm but on top of that, the Ertl algorithm also adds pseudo atoms and explicit hydrogen atoms as functional group environments which could not be represented nicely in this return type. The connected environmental carbon atoms we attach to the functional groups are also not the original carbon atoms from the molecule but substitutes that mainly contain the information whether the environmental carbon is aliphatic or aromatic. As for your suggestion "public List extractFunctionalGroups(IAtomContain mol)", I have renamed the central find() method to this name ([e6653ba](https://github.com/cdk/cdk/commit/e6653ba7bf797dce0fe36c609ca6a62c87899c71)). The method already had the same parameter and return type signature.
> 
> And about adjacency list vs atom container 2 (sorry for the long text!): thank you for pointing out that AtomContainer2 should be as fast as our solution now. I will keep this in mind. But please forgive me that I am unwilling to change this here now since this is not my code originally (I only took over the responsibility for it) and I don't want to change the entire internal implementation at this step anymore. Sorry!

No worries, I will take a look at both of these.

JonasSchaub commented:
> Utilities yes - but in the right place. I understand the CDK API is quite bloated so it can be hard to find the right thing but adding more utility methods to places that someone would not think to look doesn't help IMO. For example we already have an isMetal as part of the Elements enum class: https://github.com/cdk/cdk/blob/main/base/core/src/main/java/org/openscience/cdk/config/Elements.java#L516C27-L516C34
> 
> Would someone looking for an isMetal() etc function expect to find it on this functional group extraction algorithm?

You're absolutely right, someone searching for such utilities in general would not look for them here. Here, they would be intended as utilities _to be used in combination with this class_. 
But your point about Elements.isMetal() is very helpful. Sadly, I cannot simply do !isMetal() to detect nonmetal ("organic") atoms/structures because this would include metalloids and pseudo atoms. You wouldn't, by chance, allow me to add methods like "isMetalloid()", "isPseudoAtom()", and "isNonmetal()" to Elements, would you?

Do we also have an "isCharged()" utility somwhere already?

My "isStructureUnconnected()" I could actually easily drop and tell the user to use "!ConnectivityChecker.isConnected()", I will change this accordingly!

johnmay commented:
We should add an ``isMetaloid()`` to the Elements class. ``isCharged()`` is a bit wobbly IMO due to resonance/tautomers.


![image](https://github.com/cdk/cdk/assets/983232/e9add28d-d887-4b12-838a-5120aeee1b2d)

johnmay commented:
All looks good to me.
